pkg/controller/pki/pki_controller.go

```diff
@@ -0,0 +1,269 @@
+// business of reconciling the certificate objects
+type pki struct {
+	spec       netopv1.OperatorPKISpec
+	controller *certrotation.CertRotationController
+	lastErr    error
+}
+
+// newPKI creates a CertRotationController for the supplied configuration
+func newPKI(config *netopv1.OperatorPKI, clientset *kubernetes.Clientset, mgr manager.Manager) (*pki, error) {
+	spec := config.Spec
+
+	// Ugly: the existing cache + informers used as part of the controller-manager
+	// can't be used, because they're untyped. So, we need to create our own.
+	// However, this has a few advantages - namely, we're creating a namespace-scoped
+	// watch, which is much more efficient than watching all Secrets and ConfigMaps
+	// TODO: consider adding a label selector to the watch, since we can do that.
+
+	inf := informers.NewSharedInformerFactoryWithOptions(
+		clientset,
+		24*time.Hour,
+		informers.WithNamespace(config.Namespace))
+
+	cont, err := certrotation.NewCertRotationController(
+		fmt.Sprintf("%s/%s", config.Namespace, config.Name), // name, not really used
+		certrotation.SigningRotation{
+			Namespace:     config.Namespace,
+			Name:          config.Name + "-ca",
+			Validity:      10 * OneYear,
+			Refresh:       9 * OneYear,
+			Informer:      inf.Core().V1().Secrets(),
+			Lister:        inf.Core().V1().Secrets().Lister(),
+			Client:        clientset.CoreV1(),
+			EventRecorder: &loggingRecorder{},
+		},
+		certrotation.CABundleRotation{
+			Namespace:     config.Namespace,
+			Name:          config.Name + "-ca",
+			Lister:        inf.Core().V1().ConfigMaps().Lister(),
+			Informer:      inf.Core().V1().ConfigMaps(),
+			Client:        clientset.CoreV1(),
+			EventRecorder: &loggingRecorder{},
+		},
+		certrotation.TargetRotation{
+			Namespace: config.Namespace,
+			Name:      config.Name + "-cert",
+			Validity:  OneYear / 2,
+			Refresh:   OneYear / 4,
+			CertCreator: &certrotation.ServingRotation{
+				Hostnames: func() []string { return []string{spec.TargetCert.CommonName} },
+
+				// Force the certificate to also be client
+				CertificateExtensionFn: []crypto.CertificateExtensionFunc{
+					toClientCert,
+				},
+			},
+			Lister:        inf.Core().V1().Secrets().Lister(),
+			Informer:      inf.Core().V1().Secrets(),
+			Client:        clientset.CoreV1(),
+			EventRecorder: &loggingRecorder{},
+		},
+		nil, // no operatorclient needed
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	out := &pki{
+		controller: cont,
+	}
+	config.Spec.DeepCopyInto(&out.spec)
+
+	ch := make(chan struct{})
+	inf.Start(ch)
+	inf.WaitForCacheSync(ch)
+
+	return out, nil
+}
+
+// sync causes the underlying cert controller to try and reconcile
+func (p *pki) sync() error {
+	return p.controller.RunOnce()
+}
+
+// toClientCert is a certificate "decorator" that adds ClientAuth to the
+// list of ExtendedKeyUsages. This allows the generated certificate to be
+// used for both client and server auth.
```

Comment:
Oh, yeah, you didn't answer why you were doing this. Are you going to have the client and server using the same cert? Why not just make two certificates? That seems way more sane...

Reply:
for OVN, we have to use the same cert, since the databases will be either servers or clients, based on the replication state.

Also, the internal library-go code currently only allows creating a single cert. Eventually we could improve this, but it's not necessary now.